File: backend/pkg/core/agent/draft_service.go
package agent

import (
	"context"

	"backend/pkg/database/repository"

	"github.com/google/uuid"
)

// articleDraftAdapter implements ArticleDraftService using the article repository.
type articleDraftAdapter struct {
	repo repository.ArticleRepository
}

// NewArticleDraftService creates an ArticleDraftService backed by the article repository.
func NewArticleDraftService(repo repository.ArticleRepository) ArticleDraftService {
	return &articleDraftAdapter{repo: repo}
}

func (a *articleDraftAdapter) CreateDraftSnapshot(ctx context.Context, articleID uuid.UUID) (*uuid.UUID, error) {
	return a.repo.CreateDraftSnapshot(ctx, articleID)
}

func (a *articleDraftAdapter) UpdateDraftContent(ctx context.Context, articleID uuid.UUID, htmlContent string) error {
	return a.repo.UpdateDraftContent(ctx, articleID, htmlContent)
}

File: backend/src/core/copilot/draft_service.rs
use std::sync::Arc;

use async_trait::async_trait;
use uuid::Uuid;

use crate::{core::article::ArticleRepository, error::AppError};

#[async_trait]
pub trait ArticleDraftService: Send + Sync {
    async fn create_draft_snapshot(&self, article_id: Uuid) -> Result<Option<Uuid>, AppError>;
    async fn update_draft_content(&self, article_id: Uuid, content: &str) -> Result<(), AppError>;
}

pub struct ArticleDraftAdapter {
    repository: Arc<dyn ArticleRepository>,
}

impl ArticleDraftAdapter {
    pub fn new(repository: Arc<dyn ArticleRepository>) -> Self {
        Self { repository }
    }
}

#[async_trait]
impl ArticleDraftService for ArticleDraftAdapter {
    async fn create_draft_snapshot(&self, article_id: Uuid) -> Result<Option<Uuid>, AppError> {
        self.repository
            .create_draft_snapshot(article_id)
            .await
            .map(Some)
    }

    async fn update_draft_content(&self, article_id: Uuid, content: &str) -> Result<(), AppError> {
        self.repository
            .update_draft_content(article_id, content)
            .await
    }
}

#[async_trait]
impl<T> crate::core::ml::llm::DraftSaver for T
where
    T: ArticleDraftService + Send + Sync + ?Sized,
{
    async fn update_draft_content(
        &self,
        article_id: Uuid,
        markdown_content: &str,
    ) -> Result<(), AppError> {
        ArticleDraftService::update_draft_content(self, article_id, markdown_content).await
    }
}
